swarm/pss/pss.go

```diff
@@ -886,68 +886,83 @@ func (p *Pss) send(to []byte, topic Topic, msg []byte, asymmetric bool, key []by
 	return nil
 }
 
-// Forwards a pss message to the peer(s) closest to the to recipient address in the PssMsg struct
-// The recipient address can be of any length, and the byte slice will be matched to the MSB slice
-// of the peer address of the equivalent length.
-func (p *Pss) forward(msg *PssMsg) error {
-	metrics.GetOrRegisterCounter("pss.forward", nil).Inc(1)
+// tries to send a message, returns true if successful
+func trySendMsg(p *Pss, sp *network.Peer, msg *PssMsg) bool {
+	var isPssEnabled bool
+	info := sp.Info()
+	for _, capability := range info.Caps {
+		if capability == p.capstring {
+			isPssEnabled = true
+			break
+		}
+	}
+	if !isPssEnabled {
+		log.Trace("peer doesn't have matching pss capabilities, skipping", "peer", info.Name, "caps", info.Caps)
+		return false
+	}
+
+	// get the protocol peer from the forwarding peer cache
+	p.fwdPoolMu.RLock()
+	pp := p.fwdPool[sp.Info().ID]
+	p.fwdPoolMu.RUnlock()
+
+	err := pp.Send(context.TODO(), msg)
+	if err != nil {
+		metrics.GetOrRegisterCounter("pss.pp.send.error", nil).Inc(1)
+		log.Error(err.Error())
+	}
+
+	return err == nil
```

Comment:
I find this weird, why not just return the error that might actually hold some important information to the caller?

Reply:
I will handle this in subsequent typed error PR.

Reply:
in this particular case we only need to know how many messages were actually sent. message might not be sent just because pss is not enabled, which is not an error. so it makes sense to return a boolean, indicating: sent or not sent.